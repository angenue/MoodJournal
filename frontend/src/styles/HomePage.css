.home-page {
    background-color: #EFEFEF;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
    overflow-x: hidden;
    margin: 0;
    background-size: cover;
    height: 100vh;
  }

  .todays-date {
    padding-top: 20px;
    color: #7b9e87;
    font-size: 70px;
    text-align: center;
}

  .emojis {
    backface-visibility: hidden;
    padding: 12px;
    text-align: center;
  
    margin: 60px -44px -44px -44px;
    border-radius: 8px;
	position: relative;
	bottom: 40px;
  }

  @media (max-width: 1023px) {
    .emojis {
      margin: 30px -16px -16px -16px; } }

  @media (max-width: 600px) {
    .emojis {
      margin: 30px -8px -12px -8px;
      border-radius: 8px; } }

  .emoji-reaction-prompt {
      padding-bottom: 9px;
      text-align: center;
      color: #b6be9c;
      font-size: 20px;
  }
  
  .emojis .emoji-reaction {
    filter: grayscale(100%);
      background: none;
      color: inherit;
      border: none;
      padding-bottom: 20px;
      font: inherit;
      cursor: pointer;
      outline: none;
      width: 60px;
      display: inline-block;
      text-align: center;
      font-size: 40px;
      transition: transform 0.16s cubic-bezier(0.65, 0.61, 0.18, 1.8) 0.02s, filter 0.32s linear;
      cursor: pointer;
      transform-origin: 50% 60%;
  }

  .emojis .emoji-reaction:hover, .emojis .emoji-reaction:focus {
    filter: grayscale(0%);
          transform: scale(1.32);
        transition: transform .04s;
  }

  .emojis .emoji-reaction.active{
    filter: grayscale(0%);
    transform: scale(1.4);
    transition: transform .04s;
  }
  
  .editor-container {
    margin: 0 auto;
    max-width: 700px;
  }

  .custom-editor {
    padding: 20px; /* Adjust padding as needed */
    height: 500px;
    box-shadow: 2px 2px 10px hsla(0, 0%, 0%, 0.1);
    border-radius: 20px;
    overflow: auto; /* Add this line to enable scrollbar */
    background-color: #FFFFFF;
    font-size: 1em;
    line-height: 1.5em;
    box-sizing: border-box;
  }
  
  .custom-editor .DraftEditor-root {
    width: calc(100% - 20px); /* Adjust for padding */
    height: calc(100% - 20px); /* Adjust for padding */
    overflow: auto; /* Add this line to enable scrollbar */
  }
  
  

  .DraftEditor-root .public-DraftEditorPlaceholder-root {
    color: #0F0F0F; 

  }

  .editor-addons {
    display: flex;
  justify-content: space-between;
  align-items: center;
  }
  
  .word-limit {
    font-size: 15px;
    color: #0F0F0F;
  }

  .submit-button {
	background-color: #d2d0ba;
    padding: 10px 40px;
    border-radius: 5px;
    cursor: pointer;
    color: #5e747f;
    border: none;
    outline: none;
    margin: 5px;
    font-weight: normal;
  }

  .submit-button:hover {
    background-color: #E3E1D3;
  }

  .disabled-button {
    background-color: #ccc; /* Use the desired grey color */
    cursor: not-allowed;
  }

  .disabled-button:hover {
    background-color: #ccc; /* Use the desired grey color */
  }
  
  