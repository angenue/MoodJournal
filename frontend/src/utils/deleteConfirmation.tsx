import React from 'react';
import Modal from 'react-modal';
import 'bootstrap/dist/css/bootstrap.min.css';

Modal.setAppElement('#root');

interface CustomModalProps {
  isOpen: boolean;
  onRequestClose: () => void;
  onConfirm: () => void;
}

const CustomModal: React.FC<CustomModalProps> = ({ isOpen, onRequestClose, onConfirm }) => {
  return (
    <Modal
      isOpen={isOpen}
      onRequestClose={onRequestClose}
      style={{
        overlay: {
          backgroundColor: 'rgba(0, 0, 0, 0.5)',
        },
        content: {
          top: '50%',
          left: '50%',
          right: 'auto',
          bottom: 'auto',
          marginRight: '-50%',
          transform: 'translate(-50%, -50%)',
          backgroundColor: '#D2D0BA',
          border: 'none',
          padding: '20px',
          borderRadius: '8px',
          maxWidth: '400px',
          width: '100%',
          color: '#5e747f'
        },
      }}
    >
      <h2>Confirm Delete</h2>
      <p>Are you sure you want to delete this journal?</p>
      <div className="d-flex justify-content-between mt-4">
      <button className="btn btn-danger mr-2" onClick={onRequestClose}>Cancel</button>
      <button className="btn btn-primary" onClick={onConfirm}>Yes, Delete</button>    
      </div> 
    </Modal>
  );
};

export default CustomModal;
