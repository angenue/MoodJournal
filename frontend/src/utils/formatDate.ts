export function formatDate(inputDate: Date): string {
    const options: Intl.DateTimeFormatOptions = { year: 'numeric', month: 'short', day: 'numeric' };
        const formattedDate = new Date(inputDate).toLocaleDateString('en-US', options);

        const day = new Date(inputDate).getDate();
        const daySuffix = (day >= 11 && day <= 13) ? 'th' : (day % 10 === 1) ? 'st' : (day % 10 === 2) ? 'nd' : (day % 10 === 3) ? 'rd' : 'th';

        return formattedDate.replace(/\d{1,2}(st|nd|rd|th)/, day + daySuffix);
}