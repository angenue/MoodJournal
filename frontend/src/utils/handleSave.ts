import { Journal } from "../models/journal";

async function fetchData(input: RequestInfo, init?: RequestInit) {
  const response = await fetch(input, init);
  if (response.ok) {
      return response;
  } else {
    const errorBody = await response.json();
    const errorMessage = errorBody.error;
    throw Error(errorMessage);
      
  }
}

export interface journalInput {
    mood: string | null,
    journalEntry?: string,
    selectedDate?: Date,
  }

  export async function createJournal(note: journalInput): Promise<Journal> {
    const response = await fetchData("/api/journals",
        {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(note),
        });
    return response.json();
}
  