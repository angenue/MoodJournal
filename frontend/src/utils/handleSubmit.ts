interface FormData {
    mood: string | null;
    journalEntry: string;
    // Add other properties if needed
  }

const handleSubmit = async (formData: FormData) => {
    try {
      const response = await fetch('/api/submitForm', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(formData),
      });
  
      if (!response.ok) {
        throw new Error('Error submitting form');
      }
  
      const data = await response.json();
      return data;
    } catch (error) {
      console.error('Error submitting form', error);
      throw error; // Re-throw the error for the calling code to handle
    }
  };
  
  export default handleSubmit;
  