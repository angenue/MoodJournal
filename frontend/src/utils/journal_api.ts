import { Journal } from "../models/journal";

async function fetchData(input: RequestInfo, init?: RequestInit) {
  const response = await fetch(input, init);
  if (response.ok) {
      return response;
  } else {
    const errorBody = await response.json();
    const errorMessage = errorBody.error;
    throw Error(errorMessage);
    
  }
}

export async function fetchJournals(): Promise<Journal[]> {
  const response = await fetchData("api/journals", {method: "GET"});
  return await response.json();
}

export async function fetchJournalsByYear(year: number): Promise<Journal[]> {
  const response = await fetchData(`api/journals/${year}`, { method: "GET" });
  return await response.json();
}


export interface journalInput {
    mood: string | null,
    journalEntry?: string,
    selectedDate?: Date,
  }

  export async function createJournal(journal: journalInput): Promise<Journal> {
    const response = await fetchData("/api/journals",
        {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(journal),
        });
    return response.json();
}

export async function updateJournal(journalId: string, journal: journalInput): Promise<Journal> {
  const response = await fetchData("/api/journals/" + journalId,
  {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
  },
  body: JSON.stringify(journal),
  });
  return response.json();
}

export async function deleteJournal(journalId: string) {
  await fetchData("/api/journals/" + journalId, { method: "DELETE"});
}
  