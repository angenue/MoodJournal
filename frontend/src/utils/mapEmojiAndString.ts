export const mapEmojiToString = (emoji: string | null) => {
  if (!emoji) return "";
  switch (emoji) {
    case "😃":
      return "happy";
    case "😊":
      return "content";
    case "😐":
      return "neutral";
    case "😢":
      return "sad";
    case "😡":
      return "angry";
    default:
      return "";
  }
};

export const mapStringToEmoji = (emoji: string | null) => {
  if (!emoji) return "";
  switch (emoji) {
    case "happy":
      return "😃";
    case "content":
      return "😊";
    case "neutral":
      return "😐";
    case "sad":
      return "😢";
    case "angry":
      return "😡";
    default:
      return "";
  }
};

