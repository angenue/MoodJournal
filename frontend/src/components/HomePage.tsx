import React, { useState } from "react"
import {journalInput} from "../utils/journal_api";
import * as JournalsApi from "../utils/journal_api";
//import "../styles/HomePage.css";
import styles from "../styles/JournalEntry.module.css";
import { useForm, SubmitHandler } from "react-hook-form";
import { mapEmojiToString } from '../utils/mapEmojiToString';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';


interface FormData {
  mood: string | null;
  journalEntry?: string;
  selectedDate?: Date;
}

const HomePage = () => {

  const [selectedMood, setSelectedMood] = useState<string | null>(null);
  const [wordLimitExceeded, setWordLimitExceeded] = useState(false);
  const { register, handleSubmit, getValues, formState: {errors, isSubmitting }} = useForm<journalInput>();

  const handleSelectEmoji = (mood: string) => {
    setSelectedMood(mood);
  };

  const onSubmit: SubmitHandler<FormData> = async (formData) => {
    console.log("Form Data:", formData); // Check the console for this log
    try {
      formData.mood = mapEmojiToString(selectedMood); // Set the mood from selectedMood
      const journalResponse = await JournalsApi.createJournal(formData);
      console.log("Journal Response:", journalResponse); // Check the console for this log

      // Show a success toast
    toast("💗 Diary Submitted", {
      position: "top-center",
      autoClose: 3000, // Automatically close after 3 seconds
      hideProgressBar: false,
      closeOnClick: true,
      pauseOnHover: true,
      draggable: true,
      progress: undefined
    });
      
    } catch (error) {
      toast.error("Unable To Submit Diary", {
        position: "top-center",
        autoClose: 3000, // Automatically close after 3 seconds
        hideProgressBar: false,
        closeOnClick: true,
        pauseOnHover: true,
        draggable: true,
        progress: undefined
      });
      console.error(error);
      alert(error);
    }
  }

  const checkWordCount = () => {
    const journalEntry = getValues("journalEntry");
    if (journalEntry) {
      const wordCount = journalEntry.split(/\s/).length;
      setWordLimitExceeded(wordCount > 500);
    }
  };
  

  return (
    <div className="home-page">
      <ToastContainer />

      <h1 className={styles["todays-date"]}>{new Date().toDateString()}</h1>
      <div className={styles.emojis}>
        <div className={styles["emoji-reaction-prompt"]}>
          What best describes your mood today?
        </div>
        <button
          className={`${styles["emoji-reaction"]} ${selectedMood === "😃" ? styles["active"] : ""}`}
          onClick={() => handleSelectEmoji("😃")}
        >
          😃
        </button>
        <button
          className={`${styles["emoji-reaction"]} ${selectedMood === "😊" ? styles["active"] : ""}`}
          onClick={() => handleSelectEmoji("😊")}
        >
          😊
        </button>
        <button
          className={`${styles["emoji-reaction"]} ${selectedMood === "😐" ? styles["active"] : ""}`}
          onClick={() => handleSelectEmoji("😐")}
        >
          😐
        </button>
        <button
          className={`${styles["emoji-reaction"]} ${selectedMood === "😢" ? styles["active"] : ""}`}
          onClick={() => handleSelectEmoji("😢")}
        >
          😢
        </button>
        <button
          className={`${styles["emoji-reaction"]} ${selectedMood === "😡" ? styles["active"] : ""}`}
          onClick={() => handleSelectEmoji("😡")}
        >
          😡
        </button>
      </div>

      
      <form className={styles["editor-container"]} onSubmit={handleSubmit(onSubmit)}>
  <textarea
    className={styles["custom-editor"]}
    placeholder="Write your journal entry..."
    rows={10}
    {...register("journalEntry")}
    onChange={checkWordCount}
  />

  <input
    type="hidden"
    value={selectedMood || ''}
    {...register("mood")}
  />

<div className={styles["editor-addons"]}>
    <div className={styles["word-limit"]}>
      {`${getValues("journalEntry")?.split(/\s/).length ?? 0} words / 500 limit`}
    </div>

    <button
      className={styles["submit-button"]}
      disabled={wordLimitExceeded || isSubmitting}
      type="submit"
    >
      Submit Diary
    </button>
        </div>
      </form>
    </div>
  );
};

export default HomePage;
