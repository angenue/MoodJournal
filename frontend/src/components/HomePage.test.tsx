import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import HomePage from './HomePage';
import * as JournalsApi from '../utils/journal_api';
import { ToastContainer } from 'react-toastify';

const createJournal = JournalsApi.createJournal as jest.Mock;
const getLoggedInUser = JournalsApi.getLoggedInUser as jest.Mock;

// Mock the API module
jest.mock('../utils/journal_api', () => ({
  createJournal: jest.fn(),
  getLoggedInUser: jest.fn(),
}));

// Mock the ToastContainer to avoid rendering it during tests
jest.mock('react-toastify', () => ({
  ToastContainer: () => <div>ToastContainer</div>,
}));

describe('HomePage', () => {
  beforeEach(() => {
    // Mock the getLoggedInUser to return a user object
    getLoggedInUser.mockResolvedValue({
      id: '507f1f77bcf86cd799439011',
      email: 'testuser@example.com',
      // ... other user properties
    });
  });

  it('renders correctly', () => {
    const { asFragment } = render(<HomePage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('updates word count as user types', () => {
    render(<HomePage />);
    const textarea = screen.getByPlaceholderText('Write your journal entry...');
    fireEvent.change(textarea, { target: { value: 'Hello world' } });
    expect(screen.getByText('Word Count: 2/500')).toBeInTheDocument();
  });

  it('sets selected mood when emoji is clicked', () => {
    render(<HomePage />);
    const happyEmoji = screen.getByText('ðŸ˜ƒ');
    fireEvent.click(happyEmoji);
    expect(happyEmoji).toHaveClass('active');
  });

  it('submits the form with the correct data', async () => {
    // Mock the API call
    createJournal.mockResolvedValue({ data: 'ðŸ’— Diary Submitted' });

    render(<HomePage />);
    const happyEmoji = screen.getByText('ðŸ˜ƒ');
  const textarea = screen.getByPlaceholderText('Write your journal entry...') as HTMLTextAreaElement;
  const submitButton = screen.getByText('Submit Diary');

    // Simulate user interactions
    fireEvent.click(happyEmoji);
    fireEvent.change(textarea, { target: { value: 'Today was a good day' } });
  
    // Wait for state updates if necessary
    await waitFor(() => {
      expect(textarea.value).toBe('Today was a good day');
    });
  
    fireEvent.click(submitButton);

    // Wait for the API call to resolve
    await waitFor(() => expect(createJournal).toHaveBeenCalled());

    console.log(createJournal.mock.calls);

    // Check if the API was called with the correct parameters
    expect(createJournal).toHaveBeenCalledWith({
      mood: 'happy',
      journalEntry: 'Today was a good day',
      // ... any other properties that need to be submitted
    });
  });

  // Add more tests as needed...
});
