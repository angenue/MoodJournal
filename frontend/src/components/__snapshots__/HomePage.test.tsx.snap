// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage renders correctly 1`] = `
<DocumentFragment>
  <div
    class="home-page"
  >
    <div>
      ToastContainer
    </div>
    <h1
      class="todays-date"
    >
      Sun Nov 05 2023
    </h1>
    <div
      class="emojis"
    >
      <div
        class="emoji-reaction-prompt"
      >
        What best describes your mood today?
      </div>
      <button
        class="emoji-reaction "
      >
        😃
      </button>
      <button
        class="emoji-reaction "
      >
        😊
      </button>
      <button
        class="emoji-reaction "
      >
        😐
      </button>
      <button
        class="emoji-reaction "
      >
        😢
      </button>
      <button
        class="emoji-reaction "
      >
        😡
      </button>
    </div>
    <form
      class="editor-container"
    >
      <textarea
        class="custom-editor"
        name="journalEntry"
        placeholder="Write your journal entry..."
        rows="10"
      />
      <input
        name="mood"
        type="hidden"
        value=""
      />
      <div
        class="editor-addons"
      >
        <div
          class="word-limit"
        >
          Word Count: 0/500
        </div>
        <div
          class="button-container"
        >
          <button
            class="submit-button disabled-button"
            disabled=""
            title="Mood is required"
            type="submit"
          >
            Submit Diary
          </button>
          <div
            class="hint-text"
          >
            Mood is required
          </div>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
